exchanges/stream/processor.go

```diff
@@ -0,0 +1,117 @@
+package stream
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+)
+
+const (
+	// Book is a key type for websocket orderbook data processing
+	Book                     UpdateType = iota
+	defaultChannelBufferSize            = 10
+)
+
+var (
+	errMaxChanBufferSizeInvalid = errors.New("max channel buffer must be greater than 0")
+	errDataHandlerMustNotBeNil  = errors.New("data handler cannot be nil")
+	errChannelFull              = errors.New("channel full")
+	errKeyEmpty                 = errors.New("key is empty")
+	errNoFunctionalityToProcess = errors.New("no functionality to process")
+)
+
+// Processor is a stream processor that handles incoming data from a stream,
+// spawns workers for each unique key, and processes the data in a concurrent
+// manner.
+type Processor struct {
+	routes         map[Key]chan func() error
+	chanBufferSize int
+	dataHandler    chan interface{}
+	mtx            sync.RWMutex
+}
+
+// Key is a unique key for an individual type of work.
+type Key struct {
+	Type  UpdateType
+	Asset asset.Item
+	Base  *currency.Item
+	Quote *currency.Item
+}
+
+// UpdateType is a unique key for an individual type of work.
+type UpdateType uint8
+
+// NewProcessor returns a new stream processor. This handles incoming data
+// from a stream and processes it in a concurrent manner. It will spawn a
+// new worker for each unique key. If a worker already exists for a key, it
+// will use that worker. Functions are queued and processed in the order
+// they are received. Function closures are used for ease of implementation
+// with current systems.
+func NewProcessor(maxChanBuffer int, dataHandler chan interface{}) (*Processor, error) {
+	if maxChanBuffer <= 0 {
+		return nil, errMaxChanBufferSizeInvalid
+	}
+	if dataHandler == nil {
+		return nil, errDataHandlerMustNotBeNil
+	}
+	return &Processor{
+		routes:         make(map[Key]chan func() error),
+		chanBufferSize: maxChanBuffer,
+		dataHandler:    dataHandler,
+	}, nil
+}
+
+// Process spawns a new worker for a key if one does not already exist. It
+// will then queue the function to be processed. If the channel is full, it
+// will block until a slot is available. This tries to alleviate websocket
+// reader blocking issues.
+func (w *Processor) Process(key Key, fn func() error) error {
```

Comment:
While it does `process`, you are sending a function to a queue to be executed, so I think you should highlight in the naming that it isn't just executing the function you send to it immediately. It would be weird to receive the error "channel full" when I ask the websocket processor to process.

I think `QueueFunction` should work. This was mainly highlighted when looking at GateIO being `Processor.Process(...)` which looks funny. A user can tell from `Processor.QueueFunction` that it should be Processed by the Processor.

Reply:
That is a good suggestion 🖖